### ISSUE TITLE
Create a python module for each data science step

### ISSUE THREAD

**Issue by fernandojunior** [bluegardier/airbnb_prediction#3]
Avoid to use so generic names for your modules and functions such as `generate_dataframe.py` or `def generate_dataframe`. Try to use some contextualized/actionable names so that other data scientists can understand easily.

For example, you could create modules based on each step of a [CRISP](https://paulovasconcellos.com.br/crisp-dm-semma-e-kdd-conhe%C3%A7a-as-melhores-t%C3%A9cnicas-para-explora%C3%A7%C3%A3o-de-dados-560d294547d2)-DM process:

```
airbnb_predicton/
     __init__.py
     preprocess.py   # put functions you need to join, clean, transform and prepare features.
     modeling.py      # put functions you need to train the model and also perform predictions
     evaluate.py      # put functions you need to validate your model
```

**Comment by bluegardier** [comment]
pushed preprocess and modelling scripts. However, I'm kinda confused. Is evaluate.py the new data to predict?

**Comment by fernandojunior** [comment]
The point of `evaluate.py` module is to provide an unbiased evaluation of a final model fit using unseen data (dataset which was never exposed during model training).

Ideally, we should split our input data into three data sets: train, validation and test.
Given some input data set, Pycaret already splits train and validation datasets during cross-validation (k-folds=10) ([create_module](https://pycaret.org/create-model/)).

In this way, you just need to split the raw dataset `listening.csv` into train and test datasets: `data/raw/listening_train.csv` and `data/raw/listening_test.csv`

The `listening_train.csv` dataset will be used to feed `modeling.py` module to train and cross-validate a model.  While `listening_test.csv` dataset will be used to feed `evaluate.py` module to **generate** and **save** final model performance into some output file (eg a file `data/evaluation.json` containing `{"MAE": 0.012, "RMSE": 0.11}`).

To extract final model performance you can use `from pycaret.utils import check_metric` or some sklearn regression metric like   `sklearn.metrics import median_absolute_error`.

```
from pycaret.utils import check_metric

# Pycaret regression metrics: MAE, MSE, RMSE, R2, RMSLE, MAPE
check_metric(unseen_predictions['default'], unseen_predictions['Label'], metric = 'MAE')
```

Refs:
- https://towardsdatascience.com/train-validation-and-test-sets-72cb40cba9e7
- https://github.com/pycaret/pycaret/blob/master/pycaret/utils.py#L21
- https://scikit-learn.org/stable/modules/classes.html#regression-metrics
- https://github.com/pycaret/pycaret/blob/master/tutorials/Binary%20Classification%20Tutorial%20Level%20Beginner%20-%20%20CLF101.ipynb (see **13.0 Predict on unseen data**)

**Comment by fernandojunior** [comment]
TIP:

- Change `modeling.py` module ([lines](https://github.com/bluegardier/airbnb_prediction/blob/dev/airbnb_prediction/modelling.py#L38-L39)) to use [save_model and load_model](https://pycaret.org/save-model/) functions from pycaret instead of pickle lib

**Comment by fernandojunior** [comment]
TIP:
- Avoid to import all objects from a module `from pycaret.regression import *`. Use some alias `from pycaret import regression as pc` or import only needed objects `from pycaret.regression import setup, create_model, ...`

Ref:
- https://www.python.org/dev/peps/pep-0008/#imports

> Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. There is one defensible use case for a wildcard import, which is to republish an internal interface as part of a public API (for example, overwriting a pure Python implementation of an interface with the definitions from an optional accelerator module and exactly which definitions will be overwritten isn't known in advance).

**Comment by fernandojunior** [comment]
TIP:
- Move `def generate_dataframe` function from `generate_dataframe.py`  to `preprocess.py` module
- Rename `def generate_dataframe` function to a meaningful name (eg `generate_features`, `extract_features`, `get_features`, etc.)
- Remove `generate_dataframe.py` uneeded module

**Comment by bluegardier** [comment]
> TIP:
> 
> * Move `def generate_dataframe` function from `generate_dataframe.py`  to `preprocess.py` module
> * Rename `def generate_dataframe` function to a meaningful name (eg `generate_features`, `extract_features`, `get_features`, etc.)
> * Remove `generate_dataframe.py` uneeded module

About that, I'm actually improving the code using the `dev` branch instead of `main`. I'll delete the `master` branch to avoid further confusion.

**Comment by bluegardier** [comment]
> TIP:
> 
> * Avoid to import all objects from a module `from pycaret.regression import *`. Use some alias `from pycaret import regression as pc` or import only needed objects `from pycaret.regression import setup, create_model, ...`
> 
> See #7
> 
> Ref:
> 
> * https://www.python.org/dev/peps/pep-0008/#imports
> 
> > Wildcard imports (from  import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. There is one defensible use case for a wildcard import, which is to republish an internal interface as part of a public API (for example, overwriting a pure Python implementation of an interface with the definitions from an optional accelerator module and exactly which definitions will be overwritten isn't known in advance).

I was always wondering about this, however, I've never pursued the matter hahaha. It's properly done now!

**Comment by fernandojunior** [comment]
> About that, I'm actually improving the code using the `dev` branch instead of `main`. I'll delete the `master` branch to avoid further confusion.

I see! A good way to avoid this kind of confusion is to create a pull request for each code improvement (ie git issue). But feel to remove master branch or create pull requests to update your dev/master branch!

**Comment by bluegardier** [comment]
> The point of `evaluate.py` module is to provide an unbiased evaluation of a final model fit using unseen data (dataset which was never exposed during model training).
> 
> Ideally, we should split our input data into three data sets: train, validation and test.
> Given some input data set, Pycaret already splits train and validation datasets during cross-validation (k-folds=10) ([create_module](https://pycaret.org/create-model/)).
> 
> In this way, you just need to split the raw dataset `listening.csv` into train and test datasets: `data/raw/listening_train.csv` and `data/raw/listening_test.csv`
> 
> The `listening_train.csv` dataset will be used to feed `modeling.py` module to train and cross-validate a model. While `listening_test.csv` dataset will be used to feed `evaluate.py` module to **generate** and **save** final model performance into some output file (eg a file `data/evaluation.json` containing `{"MAE": 0.012, "RMSE": 0.11}`).
> 
> To extract final model performance you can use `from pycaret.utils import check_metric` or some sklearn regression metric like `sklearn.metrics import median_absolute_error`.
> 
> ```
> from pycaret.utils import check_metric
> 
> # Pycaret regression metrics: MAE, MSE, RMSE, R2, RMSLE, MAPE
> check_metric(unseen_predictions['default'], unseen_predictions['Label'], metric = 'MAE')
> ```
> 
> Refs:
> 
> * https://towardsdatascience.com/train-validation-and-test-sets-72cb40cba9e7
> * https://github.com/pycaret/pycaret/blob/master/pycaret/utils.py#L21
> * https://scikit-learn.org/stable/modules/classes.html#regression-metrics
> * https://github.com/pycaret/pycaret/blob/master/tutorials/Binary%20Classification%20Tutorial%20Level%20Beginner%20-%20%20CLF101.ipynb (see **13.0 Predict on unseen data**)

It's DONE!! Created the modules in the proper way which you specified. They now export the model's performance as a  json as well.

**Comment by bluegardier** [comment]
> > About that, I'm actually improving the code using the `dev` branch instead of `main`. I'll delete the `master` branch to avoid further confusion.
> 
> I see! A good way to avoid this kind of confusion is to create a pull request for each code improvement (ie git issue). But feel to remove master branch or create pull requests to update your dev/master branch!

I see! It'll be the next thing I'll do. Thanks a bunch

**Comment by bluegardier** [comment]
`master` branch now removed and main merged with dev.

**Comment by bluegardier** [comment]
Just noticed that I didn't created `evaluate.py`. Instead, I've created the evaluate process inside the `modelling.py` TrainerRegressor class

**Comment by bluegardier** [comment]
closed as I'll do it in the main.py module this time
